src/H5T.c

```diff
@@ -4993,28 +5043,53 @@ H5T__path_find_real(const H5T_t *src, const H5T_t *dst, const char *name, H5T_co
         (!table || (table && conv->is_app) || (table && !table->is_hard && !conv->is_app))) {
         assert(path != table);
         assert(NULL == path->conv.u.app_func);
-        if (path->src && (src_id = H5I_register(H5I_DATATYPE, H5T_copy(path->src, H5T_COPY_ALL), false)) < 0)
-            HGOTO_ERROR(H5E_DATATYPE, H5E_CANTREGISTER, NULL,
-                        "unable to register source conversion type for query");
-        if (path->dst && (dst_id = H5I_register(H5I_DATATYPE, H5T_copy(path->dst, H5T_COPY_ALL), false)) < 0)
-            HGOTO_ERROR(H5E_DATATYPE, H5E_CANTREGISTER, NULL,
-                        "unable to register destination conversion type for query");
+        if (path->src && (NULL == (tmp_stype = H5T_copy(path->src, H5T_COPY_ALL))))
+            HGOTO_ERROR(H5E_DATATYPE, H5E_CANTCOPY, NULL, "unable to copy source datatype");
+        if (path->dst && (NULL == (tmp_dtype = H5T_copy(path->dst, H5T_COPY_ALL))))
+            HGOTO_ERROR(H5E_DATATYPE, H5E_CANTCOPY, NULL, "unable to copy destination datatype");
+
         path->cdata.command = H5T_CONV_INIT;
         if (conv->is_app) {
+            if (tmp_stype && ((src_id = H5I_register(H5I_DATATYPE, tmp_stype, false)) < 0))
```

Comment:
How would tmp_stype or tmp_dtype be NULL here?  Shouldn't you always register src_id and dst_id?

Reply:
Based on the above code that was previously only copying the type and wrapping it in an ID if `path->src` and `path->dst` aren't NULL, I kept that same logic assuming that there's a case where the copying could get skipped. That said, if it made it into one of the following blocks, it would pass NULL or H5I_INVALID_HID depending on the type of conversion function, but after studying the code for a while I couldn't really determine when the case would be that `path->src` or `path->dst` are NULL, so I tried to leave the functionality the same.

Reply:
Wouldn't the IDs be INVALID then?

Reply:
Yes. That part of the code probably needs to be studied more since previously AND currently, if it hit that particular spot it would pass invalid stuff to the conversion functions.

Reply:
Sorry, re-read your earlier comment and you said it would pass INVALID. :-)

Reply:
I think the earlier version of the code was buggy and you can assume that the tmp_stype & tmp_dtype are non-NULL.

Reply:
I may leave this like it is for now. That function is fairly complicated, but it looks like the datatypes in question could be NULL for the library's internal no-op function, as well as an application-registered soft conversion no-op function. If that's the case it would cause an assertion in the library when trying to copy a NULL type, but these functions seem to happily accept NULL or H5I_INVALID_HID for the datatypes as they're just ignored by the function. It seems like a contrived case but I can't easily predict what people may do. The library's no-op conversion function should never hit that part of the code as it is currently, but an application function might.